fdbserver/worker.actor.cpp

```diff
@@ -554,21 +560,27 @@ ACTOR Future<Void> registrationClient(Reference<AsyncVar<Optional<ClusterControl
 			}
 		}
 
-		Future<RegisterWorkerReply> registrationReply =
+		state Future<RegisterWorkerReply> registrationReply =
 		    ccInterface->get().present()
 		        ? brokenPromiseToNever(ccInterface->get().get().registerWorker.getReply(request))
 		        : Never();
-		choose {
+		state double startTime = now();
+		loop choose {
 			when(RegisterWorkerReply reply = wait(registrationReply)) {
 				processClass = reply.processClass;
 				asyncPriorityInfo->set(reply.priorityInfo);
+				TraceEvent("WorkerJoiningCluster").detail("CCID", ccInterface->get().get().id());
+				break;
+			}
+			when(wait(delay(SERVER_KNOBS->JOIN_CLUSTER_WARNING_INTERVAL))) {
+				TraceEvent(SevWarn, "WorkerNotJoinedClusterForLongTime").detail("WaitTime", now() - startTime);
 			}
-			when(wait(ccInterface->onChange())) {}
-			when(wait(ddInterf->onChange())) {}
-			when(wait(rkInterf->onChange())) {}
-			when(wait(degraded->onChange())) {}
-			when(wait(FlowTransport::transport().onIncompatibleChanged())) {}
-			when(wait(issues->onChange())) {}
+			when(wait(ccInterface->onChange())) { break; }
+			when(wait(ddInterf->onChange())) { break; }
+			when(wait(rkInterf->onChange())) { break; }
+			when(wait(degraded->onChange())) { break; }
+			when(wait(FlowTransport::transport().onIncompatibleChanged())) { break; }
+			when(wait(issues->onChange())) { break; }
```

Comment:
I know that we've discuss about this earlier, and the goal for this loop is to not issue another RegisterWorkerRequest when waiting for one, and the new code doesn't change the behavior of the old code. However, when the worker doesn't have a valid connection string, it'll never have a valid ccInterface. So this loop will stuck in this state forever. Therefore, now I'm wondering if we still want to use the old code structure, so that the worker has a chance to re-evaluate the `connFile`, and join a cluster after the connection string is changed.

Probably need @xumengpanda to also take a look at this.

Reply:
I don't think we'd want to break from this loop. We expect, CC, DD, RK to change. So for example, if current a new CC is elected, we never register ourselves again with new CC? In general this actor is never expected to return.

I see `degraded->set(X)` only for `X = true` and never reverts, so perhaps fine in that case.

Reply:
> I don't think we'd want to break from this loop.

@vishesh 
This new inner loop is for logging `WorkerNotJoinedClusterForLongTime`. There is another outer loop which does what you described.

> so that the worker has a chance to re-evaluate the connFile, and join a cluster after the connection string is changed.

@halfprice 
To achieve `re-evaluate the connFile`, this is the wrong place to look at. You need to modify `monitorLeaderOneGeneration` and `tryBecomeLeaderInternal`, combined with a new actor+asyncVar monitoring the content of the file. Also you might want to discard all established connections once the connection string changed. So I doubt the ROI of that change.

Reply:
Make sense. Looking closely, the above code block that reading the connection string doesn't do much to decide where to send the request. My understanding is that updating the worker's connection string should be enough to trigger a ccInterface change, which get us out of this block. So I think the current code should be fine. Thanks for pointing that out!

Reply:
> My understanding is that updating the worker's connection string should be enough to trigger a ccInterface change,

Unfortunately this is not the case. If you take a look at the aforementioned two functions(there are two more functions `monitorProxiesOneGeneration` and `monitorLeaderRemotelyOneGeneration`), they don't abort current connections to the current list of coordinators when the file content changed. As a matter of fact they don't really care, except for one of them does log the mismatch but that's all it does about the changed file content.

The only case where they attempt to connect to a list of new coordinators, currently, is when the returned leader info has the `forward` flag set. That's when those actors die and rerun everything.

Reply:
Thanks Xin! This makes it more clear. What we knew was that we can live update a worker's connection string (without knowing how does it work in detail), but based on your description, if the coordinators that a worker connected to were gone, there is no way we can live update a worker's connection string without restarting the worker.

What we want to accomplish in this PR is (1) print a log when a worker failed to join a cluster for a long time, which I think is what the currently discussed code block is doing, and (2) log when a worker's connection string is changed (I think currently, we don't know what's the current connection string a worker is using, but only the initial connection string from ProgramStart). Can you point where is the best place to accomplish (2)?

Reply:
> What we knew was that we can live update a worker's connection string (without knowing how does it work in detail), but based on your description, if the coordinators that a worker connected to were gone, there is no way we can live update a worker's connection string without restarting the worker.

Yes. The only existing way is via `coordinators` command in `fdbcli`

> (2) log when a worker's connection string is changed

You mean when the connection string used to connect to the current cluster is different from the one in the file?

Reply:
Talked with @halfprice offline and sorted this out.